test_tiered_entries

Automated regression triage opened this issue.

- Test: `dragonfly/snapshot_test.py::test_tiered_entries[df_factory0]`
- Backend: iouring
- Commit: c53a44b628c28d6aa651c0da5d18e98ab863be1b
- Run: https://github.com/dragonflydb/dragonfly/actions/runs/30783745880
- Marker: run/30783745880/attempt/1

```
Reason: Failed: Timeout (>300.0s) from pytest-timeout.

_______________________ test_tiered_entries[df_factory0] _______________________

args = ()
kwargs = {'async_client': <redis.asyncio.client.Redis(<redis.asyncio.connection.ConnectionPool(<redis.asyncio.connection.Connection(host=localhost,port=43975,db=0,decode_responses=True)>)>)>}
coro = <coroutine object test_tiered_entries at 0x7f90f5c5e9e0>
task = <Task pending name='Task-73900' coro=<test_tiered_entries() running at /__w/dragonfly/dragonfly/tests/dragonfly/snapshot_test.py:944> wait_for=<Future pending cb=[Task.task_wakeup()]>>

    @functools.wraps(func)
    def inner(*args, **kwargs):
        coro = func(*args, **kwargs)
        if not inspect.isawaitable(coro):
            pyfuncitem.warn(
                pytest.PytestWarning(
                    f"The test {pyfuncitem} is marked with '@pytest.mark.asyncio' "
                    "but it is not an async function. "
                    "Please remove asyncio marker. "
                    "If the test is not marked explicitly, "
                    "check for global markers applied via 'pytestmark'."
                )
            )
            return
        task = asyncio.ensure_future(coro, loop=_loop)
        try:
>           _loop.run_until_complete(task)

_loop      = <_UnixSelectorEventLoop running=False closed=False debug=False>
args       = ()
coro       = <coroutine object test_tiered_entries at 0x7f90f5c5e9e0>
func       = <function test_tiered_entries at 0x7f90fdd889a0>
kwargs     = {'async_client': <redis.asyncio.client.Redis(<redis.asyncio.connection.ConnectionPool(<redis.asyncio.connection.Connection(host=localhost,port=43975,db=0,decode_responses=True)>)>)>}
pyfuncitem = <Function test_tiered_entries[df_factory0]>
task       = <Task pending name='Task-73900' coro=<test_tiered_entries() running at /__w/dragonfly/dragonfly/tests/dragonfly/snapshot_test.py:944> wait_for=<Future pending cb=[Task.task_wakeup()]>>

/usr/local/lib/python3.12/dist-packages/pytest_asyncio/plugin.py:449: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3.12/asyncio/base_events.py:674: in run_until_complete
    self.run_forever()
        future     = <Task pending name='Task-73900' coro=<test_tiered_entries() running at /__w/dragonfly/dragonfly/tests/dragonfly/snapshot_test.py:944> wait_for=<Future pending cb=[Task.task_wakeup()]>>
        new_task   = False
        self       = <_UnixSelectorEventLoop running=False closed=False debug=False>
/usr/lib/python3.12/asyncio/base_events.py:641: in run_forever
    self._run_once()
        old_agen_hooks = asyncgen_hooks(firstiter=None, finalizer=None)
        self       = <_UnixSelectorEventLoop running=False closed=False debug=False>
/usr/lib/python3.12/asyncio/base_events.py:1949: in _run_once
    event_list = self._selector.select(timeout)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        sched_count = 1
        self       = <_UnixSelectorEventLoop running=False closed=False debug=False>
        timeout    = 0.09999145299980228
        when       = 2314.674812674
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <selectors.EpollSelector object at 0x7f90f44df170>, timeout = 0.1

    def select(self, timeout=None):
        if timeout is None:
            timeout = -1
        elif timeout <= 0:
            timeout = 0
        else:
            # epoll_wait() has a resolution of 1 millisecond, round away
            # from zero to wait *at least* timeout seconds.
            timeout = math.ceil(timeout * 1e3) * 1e-3
    
        # epoll_wait() expects `maxevents` to be greater than zero;
        # we want to make sure that `select()` can be called when no
        # FD is registered.
        max_ev = max(len(self._fd_to_key), 1)
    
        ready = []
        try:
>           fd_event_list = self._selector.poll(timeout, max_ev)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           Failed: Timeout (>300.0s) from pytest-timeout.

```

--- a/src/server/tiering/small_bins.cc
+++ b/src/server/tiering/small_bins.cc
@@ -157,7 +157,7 @@ SmallBins::BinInfo SmallBins::Delete(DiskSegment segment) {
       return {full_segment, false /* fragmented */, true /* empty */};
     }
 
-    bool fragmented = bin.bytes < bin.orig_bytes / 2;
+    bool fragmented = bin.bytes * 2 < bin.orig_bytes;
     return {full_segment, fragmented, false /* empty */};
   }
 
@@ -166,14 +166,14 @@ SmallBins::BinInfo SmallBins::Delete(DiskSegment segment) {
 
 bool SmallBins::IsFragmented(size_t offset) {
   if (auto it = stashed_bins_.Find(offset); it != stashed_bins_.end())
-    return it->second.bytes < it->second.orig_bytes / 2;
+    return it->second.bytes * 2 < it->second.orig_bytes;
   return false;
 }
 
 ::dfly::detail::DashCursor SmallBins::TraverseFragmented(::dfly::detail::DashCursor cursor,
                                                          absl::FunctionRef<void(size_t)> f) {
   return stashed_bins_.Traverse(cursor, [f](Dash::iterator it) {
-    if (it->second.bytes < it->second.orig_bytes / 2)
+    if (it->second.bytes * 2 < it->second.orig_bytes)
       f(it->first);
   });
 }

--- a/src/server/tiering/small_bins.cc
+++ b/src/server/tiering/small_bins.cc
@@ -108,7 +108,7 @@ SmallBins::KeySegmentList SmallBins::ReportStashed(BinId id, DiskSegment segment
   }
 
   stats_.stashed_entries_cnt += list.size();
-  stashed_bins_.InsertNew(segment.offset, StashInfo{uint8_t(list.size()), bytes});
+  stashed_bins_.InsertNew(segment.offset, StashInfo{uint8_t(list.size()), bytes, bytes});
 
   return list;
 }
@@ -157,9 +157,8 @@ SmallBins::BinInfo SmallBins::Delete(DiskSegment segment) {
       return {full_segment, false /* fragmented */, true /* empty */};
     }
 
-    if (bin.bytes < kPageSize / 2) {
-      return {full_segment, true /* fragmented */, false /* empty */};
-    }
+    bool fragmented = bin.bytes < bin.orig_bytes / 2;
+    return {full_segment, fragmented, false /* empty */};
   }
 
   return {segment};
@@ -167,14 +166,14 @@ SmallBins::BinInfo SmallBins::Delete(DiskSegment segment) {
 
 bool SmallBins::IsFragmented(size_t offset) {
   if (auto it = stashed_bins_.Find(offset); it != stashed_bins_.end())
-    return it->second.bytes < kPageSize / 2;
+    return it->second.bytes < it->second.orig_bytes / 2;
   return false;
 }
 
 ::dfly::detail::DashCursor SmallBins::TraverseFragmented(::dfly::detail::DashCursor cursor,
                                                          absl::FunctionRef<void(size_t)> f) {
   return stashed_bins_.Traverse(cursor, [f](Dash::iterator it) {
-    if (it->second.bytes < kPageSize / 2)
+    if (it->second.bytes < it->second.orig_bytes / 2)
       f(it->first);
   });
 }

--- a/src/server/tiering/small_bins.h
+++ b/src/server/tiering/small_bins.h
@@ -98,9 +98,13 @@ class SmallBins {
  private:
   struct StashInfo {
     uint8_t entries = 0;
+    // Live bytes still occupied by non-deleted entries.
     uint16_t bytes = 0;
+    // Bytes at stash time; used to detect real fragmentation from deletes rather than
+    // flagging bins that were simply packed small.
+    uint16_t orig_bytes = 0;
   };
-  static_assert(sizeof(StashInfo) == sizeof(unsigned));
+  static_assert(sizeof(StashInfo) == 6);
 
   BinId last_bin_id_ = 0;
   FilledBin current_bin_{last_bin_id_};

--- a/src/server/tiering/small_bins_test.cc
+++ b/src/server/tiering/small_bins_test.cc
@@ -93,10 +93,17 @@ TEST_F(SmallBinsTest, PartialStashDelete) {
     EXPECT_EQ(key, "k"s + data.substr(segment.offset, segment.length).substr(1));
   }
 
-  // Delete all stashed values
+  // Delete all stashed values, tracking live bytes to know when we've crossed the
+  // "more than half of this bin's live content was deleted" fragmentation threshold.
+  size_t orig_bytes = 0;
+  for (auto& [dbid, key, segment] : segments)
+    orig_bytes += segment.length;
+
+  size_t deleted_bytes = 0;
   while (!segments.empty()) {
     auto segment = std::get<2>(segments.back());
     segments.pop_back();
+    deleted_bytes += segment.length;
     auto bin = bins_.Delete(segment);
 
     EXPECT_EQ(bin.segment.offset, 0u);
@@ -104,8 +111,10 @@ TEST_F(SmallBinsTest, PartialStashDelete) {
 
     if (segments.empty()) {
       EXPECT_TRUE(bin.empty);
+    } else if (deleted_bytes * 2 > orig_bytes) {
+      EXPECT_TRUE(bin.fragmented);  // more than half of the bin's live bytes were deleted
     } else {
-      EXPECT_TRUE(bin.fragmented);  // half of the values were deleted
+      EXPECT_FALSE(bin.fragmented);
     }
   }
 }

